common/src/main/java/io/netty/util/internal/PlatformDependent.java

```diff
@@ -471,6 +471,18 @@ public static LongCounter newLongCounter() {
         return new ConcurrentHashMap<K, V>(map);
     }
 
+    public static void loadFence() {
```

Comment:
The nice part of the Java Memory Model is that you don't forcibly need these to be exposed with Unsafe but you can provide a variant using a dummy atomic long field updater (to be load/stored or using a lazy set to get release semantic) that is portable and won't need Unsafe at all

Reply:
I'm not so familiar with the java memory model, how does memory reordering work in java? do you have a small example?

Reply:
that means when I use atomic -> it's like a memory barrier acquire or release right?

Reply:
Atomic set -> full fence (seq cst barrier)
Atomic get -> load fence (acquire barrier)
Atomic lazy set -> store fence (release barrier)

Reply:
We don't specifically need the barriers (since we can `getVolatile` or `putVolatile`) but we do need to use Unsafe, because we are manipulating the rings that reside in native memory. To avoid Unsafe in this case, we'll need to use VarHandles, but 4.1 (that this PR aimed at) is baselined on Java 7.